decidim-core/app/controllers/decidim/application_controller.rb

```diff
@@ -20,13 +20,29 @@ class ApplicationController < ::DecidimController
     helper Decidim::FeaturePathHelper
     helper Decidim::ViewHooksHelper
 
+    # Saves the location before loading each page so we can return to the
+    # right page.
+    before_action :store_current_location
+
     protect_from_forgery with: :exception, prepend: true
     after_action :add_vary_header
 
     layout "layouts/decidim/application"
 
     private
 
+    # Stores the url where the user will be redirected after login.
+    #
+    # Uses the `redirect_url` param or the current url if there's no param.
+    # In Devise controllers we only store the URL if it's from the params, we don't
+    # want to overwrite the stored URL for a Devise one.
+    def store_current_location
```

Comment:
What about checking the format of the request before? Something like `return unless request.format.html?`? This would save us a lot of trouble with XHR requests.